src/support/generation.c

```diff
@@ -95,196 +95,241 @@ __wt_gen_next_drain(WT_SESSION_IMPL *session, int which)
+                    WT_SET_VERBOSE_LEVEL(
+                      cookie->original_session, WT_VERB_EVICT_STUCK, WT_VERBOSE_DEBUG_1);
+                } else if (cookie->base.which == WT_GEN_CHECKPOINT) {
+                    WT_SET_VERBOSE_LEVEL(
+                      cookie->original_session, WT_VERB_CHECKPOINT, WT_VERBOSE_DEBUG_1);
+                    WT_SET_VERBOSE_LEVEL(
+                      cookie->original_session, WT_VERB_CHECKPOINT_CLEANUP, WT_VERBOSE_DEBUG_1);
+                    WT_SET_VERBOSE_LEVEL(
+                      cookie->original_session, WT_VERB_CHECKPOINT_PROGRESS, WT_VERBOSE_DEBUG_1);
                 }
+                cookie->verbose_timeout_flags = true;
+                /* Now we have enabled more logs, spin another time to get some information. */
+                continue;
+            }
 #endif
-                if (time_diff_ms >= conn->gen_drain_timeout_ms) {
-                    __wt_verbose_error(session, WT_VERB_GENERATION, "%s generation drain timed out",
-                      __gen_name(which));
-                    WT_ASSERT(session, false);
-                }
+            if (time_diff_ms >= conn->gen_drain_timeout_ms) {
+                __wt_verbose_error(cookie->original_session, WT_VERB_GENERATION,
+                  "%s generation drain timed out", __gen_name(cookie->base.which));
+                WT_ASSERT(cookie->original_session, false);
             }
         }
     }
 }
 
+/*
+ * __wt_gen_drain --
+ *     Wait for the resource to drain.
+ */
+void
+__wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
+{
+    WT_GENERATION_DRAIN_COOKIE cookie;
+
+    WT_CLEAR(cookie);
+    cookie.base.which = which;
+    cookie.base.target_generation = generation;
+    cookie.original_session = session;
+
+    __wt_epoch(session, &cookie.start);
+    __wt_session_array_walk(session, __gen_drain_callback, false, &cookie);
+}
+
+/*
+ * __gen_oldest_callback --
+ *     Check a single session's generation to find the oldest, callback from the session array walk.
+ */
+static void
+__gen_oldest_callback(WT_SESSION_IMPL *session, bool *exit_walkp, void *cookiep)
+{
+    WT_GENERATION_COOKIE *cookie;
+    uint64_t *oldest;
+    uint64_t v;
+
+    WT_UNUSED(exit_walkp);
+    cookie = (WT_GENERATION_COOKIE *)cookiep;
+    oldest = (uint64_t *)cookie->ret_arg;
+
+    WT_ORDERED_READ(v, session->generations[cookie->which]);
+    if (v != 0 && v < *oldest) {
+        *oldest = v;
+    }
+}
+
 /*
  * __gen_oldest --
  *     Return the oldest generation in use for the resource.
  */
 static uint64_t
 __gen_oldest(WT_SESSION_IMPL *session, int which)
 {
-    WT_CONNECTION_IMPL *conn;
-    WT_SESSION_IMPL *s;
-    uint64_t oldest, v;
-    uint32_t i, session_cnt;
+    WT_GENERATION_COOKIE cookie;
+    uint64_t oldest;
 
-    conn = S2C(session);
+    WT_CLEAR(cookie);
+    cookie.which = which;
+    cookie.ret_arg = &oldest;
```

Comment:
I can see what we're shooting for with the generic `ret_arg` type, but I think the indirection hurts readability too much. It'd be clearer to have `oldest_gen` and `active` fields in the cookie.